lib/Core/SQLiteBuildDB.cpp

```diff
@@ -626,25 +623,42 @@ class SQLiteBuildDB : public BuildDB {
     return DBKeyID(sqlite3_last_insert_rowid(db));
   }
 
-  KeyType getKeyForID(DBKeyID keyID) {
-    int result;
+  std::unordered_map<uint64_t, KeyID> engineKeyIDs;
 
-    // Seach for the key.
-    result = sqlite3_reset(findKeyNameForKeyIDStmt);
-    assert(result == SQLITE_OK);
-    result = sqlite3_clear_bindings(findKeyNameForKeyIDStmt);
-    assert(result == SQLITE_OK);
-    result = sqlite3_bind_int64(findKeyNameForKeyIDStmt, /*index=*/1, keyID.value);
-    assert(result == SQLITE_OK);
+  /// Maps a DBKeyID into an engine KeyID
+  KeyID getKeyIDForID(DBKeyID keyID) {
+    // Search local db <-> engine mapping cache
+    auto it = engineKeyIDs.find(keyID.value);
+    if (it == engineKeyIDs.end()) {
```

Comment:
I would invert the sense of this, so that the early exit is the fast return and most of the code can be de-dented.